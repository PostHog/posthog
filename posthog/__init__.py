import sys
from unittest.mock import MagicMock
import deltalake.writer

# DLT imports the missing function `try_get_deltatable` from `deltalake` which
# it doesn't actually use but also no longer exists. So we can safely mock
# it to allow the app to startup without module errors
deltalake.writer.try_get_deltatable = MagicMock()  # type: ignore
sys.modules["deltalake.writer"] = deltalake.writer


# This will make sure the app is always imported when
# Django starts so that shared_task will use this app.
from posthog.celery import app as celery_app  # noqa: E402

__all__ = ("celery_app",)

# snowflake-connector-python tries to access a root folder which errors out in pods.
# This sets the snowflake home directory to a relative folder
import os  # noqa: E402

os.environ["SNOWFLAKE_HOME"] = "./.snowflake"
