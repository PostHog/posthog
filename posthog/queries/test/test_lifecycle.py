import json
from unittest.mock import patch

from freezegun import freeze_time
from rest_framework.test import APIRequestFactory

from posthog.constants import FILTER_TEST_ACCOUNTS, TRENDS_LIFECYCLE
from posthog.models import Filter
from posthog.test.base import APIBaseTest, snapshot_clickhouse_queries
from posthog.utils import relative_date_parse


# parameterize tests to reuse in EE
def lifecycle_test_factory(trends, event_factory, person_factory, action_factory):
    class TestLifecycle(APIBaseTest):
        def _create_events(self, data):
            person_result = []
            for id, timestamps in data:
                with freeze_time(timestamps[0]):
                    person_result.append(
                        person_factory(
                            team_id=self.team.pk,
                            distinct_ids=[id],
                            properties={"name": id, **({"email": "test@posthog.com"} if id == "p1" else {})},
                        ),
                    )
                for timestamp in timestamps:
                    event_factory(
                        team=self.team, event="$pageview", distinct_id=id, timestamp=timestamp,
                    )
            return person_result

        def test_lifecycle_trend(self):
            self._create_events(
                data=[
                    (
                        "p1",
                        [
                            "2020-01-11T12:00:00Z",
                            "2020-01-12T12:00:00Z",
                            "2020-01-13T12:00:00Z",
                            "2020-01-15T12:00:00Z",
                            "2020-01-17T12:00:00Z",
                            "2020-01-19T12:00:00Z",
                        ],
                    ),
                    ("p2", ["2020-01-09T12:00:00Z", "2020-01-12T12:00:00Z"]),
                    ("p3", ["2020-01-12T12:00:00Z"]),
                    ("p4", ["2020-01-15T12:00:00Z"]),
                ]
            )

            result = trends().run(
                Filter(
                    data={
                        "date_from": "2020-01-12T00:00:00Z",
                        "date_to": "2020-01-19T00:00:00Z",
                        "events": [{"id": "$pageview", "type": "events", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                    }
                ),
                self.team,
            )

            self.assertLifecycleResults(
                result,
                [
                    {"status": "dormant", "data": [0, -2, -1, 0, -2, 0, -1, 0]},
                    {"status": "new", "data": [1, 0, 0, 1, 0, 0, 0, 0]},
                    {"status": "resurrecting", "data": [1, 0, 0, 1, 0, 1, 0, 1]},
                    {"status": "returning", "data": [1, 1, 0, 0, 0, 0, 0, 0]},
                ],
            )

        def test_lifecycle_trend_prop_filtering(self):

            p1 = person_factory(team_id=self.team.pk, distinct_ids=["p1"], properties={"name": "p1"})
            event_factory(
                team=self.team,
                event="$pageview",
                distinct_id="p1",
                timestamp="2020-01-11T12:00:00Z",
                properties={"$number": 1},
            )
            event_factory(
                team=self.team,
                event="$pageview",
                distinct_id="p1",
                timestamp="2020-01-12T12:00:00Z",
                properties={"$number": 1},
            )
            event_factory(
                team=self.team,
                event="$pageview",
                distinct_id="p1",
                timestamp="2020-01-13T12:00:00Z",
                properties={"$number": 1},
            )

            event_factory(
                team=self.team,
                event="$pageview",
                distinct_id="p1",
                timestamp="2020-01-15T12:00:00Z",
                properties={"$number": 1},
            )

            event_factory(
                team=self.team,
                event="$pageview",
                distinct_id="p1",
                timestamp="2020-01-17T12:00:00Z",
                properties={"$number": 1},
            )

            event_factory(
                team=self.team,
                event="$pageview",
                distinct_id="p1",
                timestamp="2020-01-19T12:00:00Z",
                properties={"$number": 1},
            )

            p2 = person_factory(team_id=self.team.pk, distinct_ids=["p2"], properties={"name": "p2"})
            event_factory(
                team=self.team, event="$pageview", distinct_id="p2", timestamp="2020-01-09T12:00:00Z",
            )
            event_factory(
                team=self.team, event="$pageview", distinct_id="p2", timestamp="2020-01-12T12:00:00Z",
            )

            p3 = person_factory(team_id=self.team.pk, distinct_ids=["p3"], properties={"name": "p3"})
            event_factory(
                team=self.team, event="$pageview", distinct_id="p3", timestamp="2020-01-12T12:00:00Z",
            )

            p4 = person_factory(team_id=self.team.pk, distinct_ids=["p4"], properties={"name": "p4"})
            event_factory(
                team=self.team, event="$pageview", distinct_id="p4", timestamp="2020-01-15T12:00:00Z",
            )

            result = trends().run(
                Filter(
                    data={
                        "date_from": "2020-01-12T00:00:00Z",
                        "date_to": "2020-01-19T00:00:00Z",
                        "events": [{"id": "$pageview", "type": "events", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                        "properties": [{"key": "$number", "value": 1}],
                    }
                ),
                self.team,
            )

            self.assertLifecycleResults(
                result,
                [
                    {"status": "dormant", "data": [0, 0, -1, 0, -1, 0, -1, 0]},
                    {"status": "new", "data": [0, 0, 0, 0, 0, 0, 0, 0]},
                    {"status": "resurrecting", "data": [0, 0, 0, 1, 0, 1, 0, 1]},
                    {"status": "returning", "data": [1, 1, 0, 0, 0, 0, 0, 0]},
                ],
            )

        def test_lifecycle_trends_distinct_id_repeat(self):
            with freeze_time("2020-01-12T12:00:00Z"):
                p1 = person_factory(team_id=self.team.pk, distinct_ids=["p1", "another_p1"], properties={"name": "p1"})

            event_factory(
                team=self.team, event="$pageview", distinct_id="p1", timestamp="2020-01-12T12:00:00Z",
            )
            event_factory(
                team=self.team, event="$pageview", distinct_id="another_p1", timestamp="2020-01-14T12:00:00Z",
            )
            event_factory(
                team=self.team, event="$pageview", distinct_id="p1", timestamp="2020-01-15T12:00:00Z",
            )

            event_factory(
                team=self.team, event="$pageview", distinct_id="p1", timestamp="2020-01-17T12:00:00Z",
            )

            event_factory(
                team=self.team, event="$pageview", distinct_id="p1", timestamp="2020-01-19T12:00:00Z",
            )

            result = trends().run(
                Filter(
                    data={
                        "date_from": "2020-01-12T00:00:00Z",
                        "date_to": "2020-01-19T00:00:00Z",
                        "events": [{"id": "$pageview", "type": "events", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                    }
                ),
                self.team,
            )

            self.assertLifecycleResults(
                result,
                [
                    {"status": "dormant", "data": [0, -1, 0, 0, -1, 0, -1, 0]},
                    {"status": "new", "data": [1, 0, 0, 0, 0, 0, 0, 0]},
                    {"status": "resurrecting", "data": [0, 0, 1, 0, 0, 1, 0, 1]},
                    {"status": "returning", "data": [0, 0, 0, 1, 0, 0, 0, 0]},
                ],
            )

        def test_lifecycle_trend_people(self):

            people = self._create_events(
                data=[
                    (
                        "p1",
                        [
                            "2020-01-11T12:00:00Z",
                            "2020-01-12T12:00:00Z",
                            "2020-01-13T12:00:00Z",
                            "2020-01-15T12:00:00Z",
                            "2020-01-17T12:00:00Z",
                            "2020-01-19T12:00:00Z",
                        ],
                    ),
                    ("p2", ["2020-01-09T12:00:00Z", "2020-01-12T12:00:00Z"]),
                    ("p3", ["2020-01-12T12:00:00Z"]),
                    ("p4", ["2020-01-15T12:00:00Z"]),
                ]
            )

            p1 = people[0]
            request_factory = APIRequestFactory()
            request = request_factory.get("/person/lifecycle")

            result = trends().get_people(
                Filter(
                    data={
                        "date_from": "2020-01-12T00:00:00Z",
                        "date_to": "2020-01-19T00:00:00Z",
                        "events": [{"id": "$pageview", "type": "events", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                    }
                ),
                self.team,
                relative_date_parse("2020-01-13T00:00:00Z"),
                "returning",
                request,
            )

            self.assertEqual(len(result), 1)
            self.assertEqual(result[0]["id"], p1.pk)

            dormant_result = trends().get_people(
                Filter(
                    data={
                        "date_from": "2020-01-12T00:00:00Z",
                        "date_to": "2020-01-19T00:00:00Z",
                        "events": [{"id": "$pageview", "type": "events", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                    }
                ),
                self.team,
                relative_date_parse("2020-01-13T00:00:00Z"),
                "dormant",
                request,
            )

            self.assertEqual(len(dormant_result), 2)

            dormant_result = trends().get_people(
                Filter(
                    data={
                        "date_from": "2020-01-12T00:00:00Z",
                        "date_to": "2020-01-19T00:00:00Z",
                        "events": [{"id": "$pageview", "type": "events", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                    }
                ),
                self.team,
                relative_date_parse("2020-01-14T00:00:00Z"),
                "dormant",
                request,
            )

            self.assertEqual(len(dormant_result), 1)

        def test_lifecycle_trend_people_paginated(self):
            with freeze_time("2020-01-15T12:00:00Z"):
                for i in range(150):
                    person_id = "person{}".format(i)
                    person_factory(team_id=self.team.pk, distinct_ids=[person_id])
                    event_factory(
                        team=self.team, event="$pageview", distinct_id=person_id, timestamp="2020-01-15T12:00:00Z",
                    )
            # even if set to hour 6 it should default to beginning of day and include all pageviews above
            result = self.client.get(
                "/api/person/lifecycle",
                data={
                    "date_from": "2020-01-12T00:00:00Z",
                    "date_to": "2020-01-19T00:00:00Z",
                    "events": json.dumps([{"id": "$pageview", "type": "events", "order": 0}]),
                    "shown_as": TRENDS_LIFECYCLE,
                    "lifecycle_type": "new",
                    "target_date": "2020-01-15T00:00:00Z",
                },
            ).json()
            self.assertEqual(len(result["results"][0]["people"]), 100)

            second_result = self.client.get(result["next"]).json()
            self.assertEqual(len(second_result["results"][0]["people"]), 50)

        def test_lifecycle_trend_action(self):

            self._create_events(
                data=[
                    (
                        "p1",
                        [
                            "2020-01-11T12:00:00Z",
                            "2020-01-12T12:00:00Z",
                            "2020-01-13T12:00:00Z",
                            "2020-01-15T12:00:00Z",
                            "2020-01-17T12:00:00Z",
                            "2020-01-19T12:00:00Z",
                        ],
                    ),
                    ("p2", ["2020-01-09T12:00:00Z", "2020-01-12T12:00:00Z"]),
                    ("p3", ["2020-01-12T12:00:00Z"]),
                    ("p4", ["2020-01-15T12:00:00Z"]),
                ]
            )

            pageview_action = action_factory(team=self.team, name="$pageview")

            result = trends().run(
                Filter(
                    data={
                        "date_from": "2020-01-12T00:00:00Z",
                        "date_to": "2020-01-19T00:00:00Z",
                        "actions": [{"id": pageview_action.pk, "type": "actions", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                    }
                ),
                self.team,
            )

            self.assertLifecycleResults(
                result,
                [
                    {"status": "dormant", "data": [0, -2, -1, 0, -2, 0, -1, 0]},
                    {"status": "new", "data": [1, 0, 0, 1, 0, 0, 0, 0]},
                    {"status": "resurrecting", "data": [1, 0, 0, 1, 0, 1, 0, 1]},
                    {"status": "returning", "data": [1, 1, 0, 0, 0, 0, 0, 0]},
                ],
            )

        def test_lifecycle_trend_all_time(self):
            self._create_events(
                data=[
                    (
                        "p1",
                        [
                            "2020-01-11T12:00:00Z",
                            "2020-01-12T12:00:00Z",
                            "2020-01-13T12:00:00Z",
                            "2020-01-15T12:00:00Z",
                            "2020-01-17T12:00:00Z",
                            "2020-01-19T12:00:00Z",
                        ],
                    ),
                    ("p2", ["2020-01-09T12:00:00Z", "2020-01-12T12:00:00Z"]),
                    ("p3", ["2020-01-12T12:00:00Z"]),
                    ("p4", ["2020-01-15T12:00:00Z"]),
                ]
            )

            with freeze_time("2020-01-17T13:01:01Z"):
                result = trends().run(
                    Filter(
                        data={
                            "date_from": "all",
                            "events": [{"id": "$pageview", "type": "events", "order": 0}],
                            "shown_as": TRENDS_LIFECYCLE,
                        }
                    ),
                    self.team,
                )

            self.assertLifecycleResults(
                result,
                [
                    {"status": "dormant", "data": [0, -1, 0, 0, -2, -1, 0, -2, 0]},
                    {"status": "new", "data": [1, 0, 1, 1, 0, 0, 1, 0, 0]},
                    {"status": "returning", "data": [0, 0, 0, 1, 1, 0, 0, 0, 0]},
                    {"status": "resurrecting", "data": [0, 0, 0, 1, 0, 0, 1, 0, 1]},
                ],
            )

        def test_lifecycle_trend_weeks(self):
            # lifecycle weeks rounds the date to the nearest following week  2/5 -> 2/10
            self._create_events(
                data=[
                    (
                        "p1",
                        [
                            "2020-02-01T12:00:00Z",
                            "2020-02-05T12:00:00Z",
                            "2020-02-10T12:00:00Z",
                            "2020-02-15T12:00:00Z",
                            "2020-02-27T12:00:00Z",
                            "2020-03-02T12:00:00Z",
                        ],
                    ),
                    ("p2", ["2020-02-11T12:00:00Z", "2020-02-18T12:00:00Z"]),
                    ("p3", ["2020-02-12T12:00:00Z"]),
                    ("p4", ["2020-02-27T12:00:00Z"]),
                ]
            )

            result = trends().run(
                Filter(
                    data={
                        "date_from": "2020-02-05T00:00:00Z",
                        "date_to": "2020-03-09T00:00:00Z",
                        "events": [{"id": "$pageview", "type": "events", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                        "interval": "week",
                    }
                ),
                self.team,
            )

            self.assertEqual(
                result[0]["days"], ["2020-02-03", "2020-02-10", "2020-02-17", "2020-02-24", "2020-03-02", "2020-03-09"]
            )

            self.assertLifecycleResults(
                result,
                [
                    {"status": "dormant", "data": [0, 0, -2, -1, -1, -1]},
                    {"status": "new", "data": [0, 2, 0, 1, 0, 0]},
                    {"status": "resurrecting", "data": [0, 0, 0, 1, 0, 0]},
                    {"status": "returning", "data": [1, 1, 1, 0, 1, 0]},
                ],
            )

        def test_lifecycle_trend_months(self):
            self._create_events(
                data=[
                    (
                        "p1",
                        [
                            "2020-01-11T12:00:00Z",
                            "2020-02-12T12:00:00Z",
                            "2020-03-13T12:00:00Z",
                            "2020-05-15T12:00:00Z",
                            "2020-07-17T12:00:00Z",
                            "2020-09-19T12:00:00Z",
                        ],
                    ),
                    ("p2", ["2019-12-09T12:00:00Z", "2020-02-12T12:00:00Z",]),
                    ("p3", ["2020-02-12T12:00:00Z"]),
                    ("p4", ["2020-05-15T12:00:00Z"]),
                ]
            )

            result = trends().run(
                Filter(
                    data={
                        "date_from": "2020-02-01T00:00:00Z",
                        "date_to": "2020-09-01T00:00:00Z",
                        "events": [{"id": "$pageview", "type": "events", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                        "interval": "month",
                    }
                ),
                self.team,
            )

            self.assertLifecycleResults(
                result,
                [
                    {"status": "dormant", "data": [0, -2, -1, 0, -2, 0, -1, 0]},
                    {"status": "new", "data": [1, 0, 0, 1, 0, 0, 0, 0]},
                    {"status": "resurrecting", "data": [1, 0, 0, 1, 0, 1, 0, 1]},
                    {"status": "returning", "data": [1, 1, 0, 0, 0, 0, 0, 0]},
                ],
            )

        def test_filter_test_accounts(self):
            self._create_events(
                data=[
                    (
                        "p1",  # p1 gets test@posthog.com as email and gets filtered out
                        [
                            "2020-01-11T12:00:00Z",
                            "2020-01-12T12:00:00Z",
                            "2020-01-13T12:00:00Z",
                            "2020-01-15T12:00:00Z",
                            "2020-01-17T12:00:00Z",
                            "2020-01-19T12:00:00Z",
                        ],
                    ),
                    ("p2", ["2020-01-09T12:00:00Z", "2020-01-12T12:00:00Z"]),
                    ("p3", ["2020-01-12T12:00:00Z"]),
                    ("p4", ["2020-01-15T12:00:00Z"]),
                ]
            )

            result = trends().run(
                Filter(
                    data={
                        "date_from": "2020-01-12T00:00:00Z",
                        "date_to": "2020-01-19T00:00:00Z",
                        "events": [{"id": "$pageview", "type": "events", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                        FILTER_TEST_ACCOUNTS: True,
                    },
                    team=self.team,
                ),
                self.team,
            )

            self.assertLifecycleResults(
                result,
                [
                    {"status": "dormant", "data": [0, -2, 0, 0, -1, 0, 0, 0]},
                    {"status": "new", "data": [1, 0, 0, 1, 0, 0, 0, 0]},
                    {"status": "resurrecting", "data": [1, 0, 0, 0, 0, 0, 0, 0]},
                    {"status": "returning", "data": [0, 0, 0, 0, 0, 0, 0, 0]},
                ],
            )

            request_factory = APIRequestFactory()
            request = request_factory.get("/person/lifecycle")

            dormant_result = trends().get_people(
                Filter(
                    data={
                        "date_from": "2020-01-12T00:00:00Z",
                        "date_to": "2020-01-19T00:00:00Z",
                        "events": [{"id": "$pageview", "type": "events", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                        FILTER_TEST_ACCOUNTS: True,
                    },
                    team=self.team,
                ),
                self.team,
                relative_date_parse("2020-01-13T00:00:00Z"),
                "dormant",
                request,
            )

        def assertLifecycleResults(self, results, expected):
            sorted_results = [
                {"status": r["status"], "data": r["data"]} for r in sorted(results, key=lambda r: r["status"])
            ]
            sorted_expected = list(sorted(expected, key=lambda r: r["status"]))

            self.assertEquals(sorted_results, sorted_expected)

        @snapshot_clickhouse_queries
        @patch("posthoganalytics.feature_enabled", return_value=True)
        def test_timezones(self, patch_something):
            self._create_events(
                data=[
                    (
                        "p1",
                        [
                            "2020-01-11T23:00:00Z",
                            "2020-01-12T01:00:00Z",
                            "2020-01-13T12:00:00Z",
                            "2020-01-15T12:00:00Z",
                            "2020-01-17T12:00:00Z",
                            "2020-01-19T12:00:00Z",
                        ],
                    ),
                    ("p2", ["2020-01-09T12:00:00Z", "2020-01-12T12:00:00Z"]),
                    ("p3", ["2020-01-12T12:00:00Z"]),
                    ("p4", ["2020-01-15T12:00:00Z"]),
                ]
            )

            result = trends().run(
                Filter(
                    data={
                        "date_from": "2020-01-12T00:00:00Z",
                        "date_to": "2020-01-19T00:00:00Z",
                        "events": [{"id": "$pageview", "type": "events", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                    }
                ),
                self.team,
            )

            self.assertLifecycleResults(
                result,
                [
                    {"status": "dormant", "data": [0, -2, -1, 0, -2, 0, -1, 0]},
                    {"status": "new", "data": [1, 0, 0, 1, 0, 0, 0, 0]},
                    {"status": "resurrecting", "data": [1, 0, 0, 1, 0, 1, 0, 1]},
                    {"status": "returning", "data": [1, 1, 0, 0, 0, 0, 0, 0]},
                ],
            )

            self.team.timezone = "US/Pacific"
            self.team.save()

            result_pacific = trends().run(
                Filter(
                    data={
                        "date_from": "2020-01-12T00:00:00Z",
                        "date_to": "2020-01-19T00:00:00Z",
                        "events": [{"id": "$pageview", "type": "events", "order": 0}],
                        "shown_as": TRENDS_LIFECYCLE,
                    },
                    team=self.team,
                ),
                self.team,
            )
            self.assertLifecycleResults(
                result_pacific,
                [
                    {"status": "dormant", "data": [-1.0, -2.0, -1.0, 0.0, -2.0, 0.0, -1.0, 0.0]},
                    {"status": "new", "data": [1, 0, 0, 1, 0, 0, 0, 0]},
                    {"status": "resurrecting", "data": [1, 1, 0, 1, 0, 1, 0, 1]},
                    {"status": "returning", "data": [0, 0, 0, 0, 0, 0, 0, 0]},
                ],
            )

    return TestLifecycle
