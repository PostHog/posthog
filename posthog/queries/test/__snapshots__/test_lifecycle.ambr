# name: TestLifecycle.test_sampling
  '
  WITH 'day' AS selected_period,
       periods AS
    (SELECT dateSub(day, number, dateTrunc(selected_period, toDateTime('2020-01-19 23:59:59', 'UTC'))) AS start_of_period
     FROM numbers(dateDiff('day', dateTrunc('day', toDateTime('2020-01-12 00:00:00', 'UTC')), dateTrunc('day', toDateTime('2020-01-19 23:59:59', 'UTC') + INTERVAL 1 day))))
  SELECT groupArray(start_of_period) AS date,
         groupArray(counts) AS total,
         status
  FROM
    (SELECT if(status = 'dormant', toInt64(SUM(counts)) * toInt16(-1), toInt64(SUM(counts))) as counts,
            start_of_period,
            status
     FROM
       (SELECT periods.start_of_period as start_of_period,
               toUInt16(0) AS counts,
               status
        FROM periods
        CROSS JOIN
          (SELECT status
           FROM
             (SELECT ['new', 'returning', 'resurrecting', 'dormant'] as status) ARRAY
           JOIN status) as sec
        ORDER BY status,
                 start_of_period
        UNION ALL SELECT start_of_period,
                         count(DISTINCT person_id) counts,
                         status
        FROM
          (SELECT pdi.person_id as person_id,
                  arraySort(groupUniqArray(dateTrunc('day', toTimeZone(toDateTime(events.timestamp, 'UTC'), 'UTC')))) AS all_activity,
                  arrayPopBack(arrayPushFront(all_activity, dateTrunc('day', toTimeZone(toDateTime(min(person.created_at), 'UTC'), 'UTC')))) as previous_activity,
                  arrayPopFront(arrayPushBack(all_activity, dateTrunc('day', toDateTime('1970-01-01')))) as following_activity,
                  arrayMap((previous, current, index) -> if(previous = current, 'new', if(current - INTERVAL 1 day = previous
                                                                                          AND index != 1, 'returning', 'resurrecting')), previous_activity, all_activity, arrayEnumerate(all_activity)) as initial_status,
                  arrayMap((current, next) -> if(current + INTERVAL 1 day = next, '', 'dormant'), all_activity, following_activity) as dormant_status,
                  arrayMap(x -> x + INTERVAL 1 day, arrayFilter((current, is_dormant) -> is_dormant = 'dormant', all_activity, dormant_status)) as dormant_periods,
                  arrayMap(x -> 'dormant', dormant_periods) as dormant_label,
                  arrayConcat(arrayZip(all_activity, initial_status), arrayZip(dormant_periods, dormant_label)) as temp_concat,
                  arrayJoin(temp_concat) as period_status_pairs,
                  period_status_pairs.1 as start_of_period,
                  period_status_pairs.2 as status,
                  toDateTime(min(person.created_at), 'UTC') AS created_at
           FROM events AS e SAMPLE 0.1
           INNER JOIN
             (SELECT distinct_id,
                     argMax(person_id, version) as person_id
              FROM person_distinct_id2
              WHERE team_id = 2
              GROUP BY distinct_id
              HAVING argMax(is_deleted, version) = 0) AS pdi ON e.distinct_id = pdi.distinct_id
           INNER JOIN
             (SELECT id,
                     argMax(created_at, version) as created_at
              FROM person
              WHERE team_id = 2
              GROUP BY id
              HAVING max(is_deleted) = 0 SETTINGS optimize_aggregation_in_order = 1) person ON person.id = pdi.person_id
           WHERE team_id = 2
             AND event = '$pageview'
             AND timestamp >= toDateTime(dateTrunc('day', toDateTime('2020-01-12 00:00:00', 'UTC'))) - INTERVAL 1 day
             AND timestamp < toDateTime(dateTrunc('day', toDateTime('2020-01-19 23:59:59', 'UTC'))) + INTERVAL 1 day
           GROUP BY pdi.person_id)
        GROUP BY start_of_period,
                 status)
     WHERE start_of_period <= dateTrunc('day', toDateTime('2020-01-19 23:59:59', 'UTC'))
       AND start_of_period >= dateTrunc('day', toDateTime('2020-01-12 00:00:00', 'UTC'))
     GROUP BY start_of_period,
              status
     ORDER BY start_of_period ASC)
  GROUP BY status
  '
---
# name: TestLifecycle.test_timezones
  '
  WITH 'day' AS selected_period,
       periods AS
    (SELECT dateSub(day, number, dateTrunc(selected_period, toDateTime('2020-01-19 23:59:59', 'UTC'))) AS start_of_period
     FROM numbers(dateDiff('day', dateTrunc('day', toDateTime('2020-01-12 00:00:00', 'UTC')), dateTrunc('day', toDateTime('2020-01-19 23:59:59', 'UTC') + INTERVAL 1 day))))
  SELECT groupArray(start_of_period) AS date,
         groupArray(counts) AS total,
         status
  FROM
    (SELECT if(status = 'dormant', toInt64(SUM(counts)) * toInt16(-1), toInt64(SUM(counts))) as counts,
            start_of_period,
            status
     FROM
       (SELECT periods.start_of_period as start_of_period,
               toUInt16(0) AS counts,
               status
        FROM periods
        CROSS JOIN
          (SELECT status
           FROM
             (SELECT ['new', 'returning', 'resurrecting', 'dormant'] as status) ARRAY
           JOIN status) as sec
        ORDER BY status,
                 start_of_period
        UNION ALL SELECT start_of_period,
                         count(DISTINCT person_id) counts,
                         status
        FROM
          (SELECT pdi.person_id as person_id,
                  arraySort(groupUniqArray(dateTrunc('day', toTimeZone(toDateTime(events.timestamp, 'UTC'), 'UTC')))) AS all_activity,
                  arrayPopBack(arrayPushFront(all_activity, dateTrunc('day', toTimeZone(toDateTime(min(person.created_at), 'UTC'), 'UTC')))) as previous_activity,
                  arrayPopFront(arrayPushBack(all_activity, dateTrunc('day', toDateTime('1970-01-01')))) as following_activity,
                  arrayMap((previous, current, index) -> if(previous = current, 'new', if(current - INTERVAL 1 day = previous
                                                                                          AND index != 1, 'returning', 'resurrecting')), previous_activity, all_activity, arrayEnumerate(all_activity)) as initial_status,
                  arrayMap((current, next) -> if(current + INTERVAL 1 day = next, '', 'dormant'), all_activity, following_activity) as dormant_status,
                  arrayMap(x -> x + INTERVAL 1 day, arrayFilter((current, is_dormant) -> is_dormant = 'dormant', all_activity, dormant_status)) as dormant_periods,
                  arrayMap(x -> 'dormant', dormant_periods) as dormant_label,
                  arrayConcat(arrayZip(all_activity, initial_status), arrayZip(dormant_periods, dormant_label)) as temp_concat,
                  arrayJoin(temp_concat) as period_status_pairs,
                  period_status_pairs.1 as start_of_period,
                  period_status_pairs.2 as status,
                  toDateTime(min(person.created_at), 'UTC') AS created_at
           FROM events AS e
           INNER JOIN
             (SELECT distinct_id,
                     argMax(person_id, version) as person_id
              FROM person_distinct_id2
              WHERE team_id = 2
              GROUP BY distinct_id
              HAVING argMax(is_deleted, version) = 0) AS pdi ON e.distinct_id = pdi.distinct_id
           INNER JOIN
             (SELECT id,
                     argMax(created_at, version) as created_at
              FROM person
              WHERE team_id = 2
              GROUP BY id
              HAVING max(is_deleted) = 0 SETTINGS optimize_aggregation_in_order = 1) person ON person.id = pdi.person_id
           WHERE team_id = 2
             AND event = '$pageview'
             AND timestamp >= toDateTime(dateTrunc('day', toDateTime('2020-01-12 00:00:00', 'UTC'))) - INTERVAL 1 day
             AND timestamp < toDateTime(dateTrunc('day', toDateTime('2020-01-19 23:59:59', 'UTC'))) + INTERVAL 1 day
           GROUP BY pdi.person_id)
        GROUP BY start_of_period,
                 status)
     WHERE start_of_period <= dateTrunc('day', toDateTime('2020-01-19 23:59:59', 'UTC'))
       AND start_of_period >= dateTrunc('day', toDateTime('2020-01-12 00:00:00', 'UTC'))
     GROUP BY start_of_period,
              status
     ORDER BY start_of_period ASC)
  GROUP BY status
  '
---
# name: TestLifecycle.test_timezones.1
  '
  WITH 'day' AS selected_period,
       periods AS
    (SELECT dateSub(day, number, dateTrunc(selected_period, toDateTime('2020-01-19 23:59:59', 'US/Pacific'))) AS start_of_period
     FROM numbers(dateDiff('day', dateTrunc('day', toDateTime('2020-01-12 00:00:00', 'US/Pacific')), dateTrunc('day', toDateTime('2020-01-19 23:59:59', 'US/Pacific') + INTERVAL 1 day))))
  SELECT groupArray(start_of_period) AS date,
         groupArray(counts) AS total,
         status
  FROM
    (SELECT if(status = 'dormant', toInt64(SUM(counts)) * toInt16(-1), toInt64(SUM(counts))) as counts,
            start_of_period,
            status
     FROM
       (SELECT periods.start_of_period as start_of_period,
               toUInt16(0) AS counts,
               status
        FROM periods
        CROSS JOIN
          (SELECT status
           FROM
             (SELECT ['new', 'returning', 'resurrecting', 'dormant'] as status) ARRAY
           JOIN status) as sec
        ORDER BY status,
                 start_of_period
        UNION ALL SELECT start_of_period,
                         count(DISTINCT person_id) counts,
                         status
        FROM
          (SELECT pdi.person_id as person_id,
                  arraySort(groupUniqArray(dateTrunc('day', toTimeZone(toDateTime(events.timestamp, 'US/Pacific'), 'US/Pacific')))) AS all_activity,
                  arrayPopBack(arrayPushFront(all_activity, dateTrunc('day', toTimeZone(toDateTime(min(person.created_at), 'US/Pacific'), 'US/Pacific')))) as previous_activity,
                  arrayPopFront(arrayPushBack(all_activity, dateTrunc('day', toDateTime('1970-01-01')))) as following_activity,
                  arrayMap((previous, current, index) -> if(previous = current, 'new', if(current - INTERVAL 1 day = previous
                                                                                          AND index != 1, 'returning', 'resurrecting')), previous_activity, all_activity, arrayEnumerate(all_activity)) as initial_status,
                  arrayMap((current, next) -> if(current + INTERVAL 1 day = next, '', 'dormant'), all_activity, following_activity) as dormant_status,
                  arrayMap(x -> x + INTERVAL 1 day, arrayFilter((current, is_dormant) -> is_dormant = 'dormant', all_activity, dormant_status)) as dormant_periods,
                  arrayMap(x -> 'dormant', dormant_periods) as dormant_label,
                  arrayConcat(arrayZip(all_activity, initial_status), arrayZip(dormant_periods, dormant_label)) as temp_concat,
                  arrayJoin(temp_concat) as period_status_pairs,
                  period_status_pairs.1 as start_of_period,
                  period_status_pairs.2 as status,
                  toDateTime(min(person.created_at), 'US/Pacific') AS created_at
           FROM events AS e
           INNER JOIN
             (SELECT distinct_id,
                     argMax(person_id, version) as person_id
              FROM person_distinct_id2
              WHERE team_id = 2
              GROUP BY distinct_id
              HAVING argMax(is_deleted, version) = 0) AS pdi ON e.distinct_id = pdi.distinct_id
           INNER JOIN
             (SELECT id,
                     argMax(created_at, version) as created_at
              FROM person
              WHERE team_id = 2
              GROUP BY id
              HAVING max(is_deleted) = 0 SETTINGS optimize_aggregation_in_order = 1) person ON person.id = pdi.person_id
           WHERE team_id = 2
             AND event = '$pageview'
             AND timestamp >= toDateTime(dateTrunc('day', toDateTime('2020-01-12 00:00:00', 'US/Pacific'))) - INTERVAL 1 day
             AND timestamp < toDateTime(dateTrunc('day', toDateTime('2020-01-19 23:59:59', 'US/Pacific'))) + INTERVAL 1 day
           GROUP BY pdi.person_id)
        GROUP BY start_of_period,
                 status)
     WHERE start_of_period <= dateTrunc('day', toDateTime('2020-01-19 23:59:59', 'US/Pacific'))
       AND start_of_period >= dateTrunc('day', toDateTime('2020-01-12 00:00:00', 'US/Pacific'))
     GROUP BY start_of_period,
              status
     ORDER BY start_of_period ASC)
  GROUP BY status
  '
---
